server/src/db/mod.rs

```diff
@@ -0,0 +1,106 @@
+pub mod models;
+pub mod schema;
+
+use anyhow::Result;
+use diesel::{pg::PgConnection, r2d2};
+use serde::Deserialize;
+
+pub type Pool = r2d2::Pool<r2d2::ConnectionManager<PgConnection>>;
+
+#[derive(Clone, Deserialize)]
+/// Configuration params for database. See https://docs.rs/postgres/0.17.3/postgres/config/struct.Config.html
+pub struct DatabaseSettings {
+    pub user: String,
+    pub password: String,
+    pub host: String,
+    pub port: u16,
+    pub dbname: String,
+    pub application_name: Option<String>,
+    pub connect_timeout_sec: u64,
+    pub pool_timeout_ms: u64,
+    pub read_timeout_ms: u64,
+}
+
+impl DatabaseSettings {
+    /// Obtain a [postgres url](https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING)
+    /// with the given configuration parameters
+    fn to_postgres_url(&self) -> String {
+        let mut url = format!(
+            "postgres://{}:{}@{}:{}/{}",
+            self.user, self.password, self.host, self.port, self.dbname
+        );
+
+        let mut params = Vec::new();
+
+        if let Some(app_name) = &self.application_name {
+            params.push(format!("application_name={}", app_name));
+        }
+
+        if self.connect_timeout_sec > 0 {
+            params.push(format!("connect_timeout={}", self.connect_timeout_sec));
+        }
+
+        if !params.is_empty() {
+            url += "?";
+            url += &params.join("&");
+        }
```

Comment:
Why += on separate lines? Do we expect this to change a lot?

Reply:
It's the result of a quirk of Rust - there are different types for a dynamic string buffer (String) vs an immutable sequence of characters (str). You can use "+" to append String + str, but not any other combination of the types. In this case url is a String, "?" is a str, and params.join("&") is a String, so the only way to do this in one line would be `url = url + "?" + &params.join("&")`, which imo is a bit less clean.